index.js

```diff
@@ -34,19 +39,210 @@ class ServerlessPlugin {
 
             return delegate(variableString);
         }
+
+        this.hooks = {
+            'package:setupProviderConfiguration': this.createVendorZip.bind(this),
+            'after:aws:deploy:deploy:createStack': this.uploadVendorZip.bind(this),
+            'before:remove:remove': this.removeVendorArchives.bind(this)
+        };
     }
 
-    checkCompatibleRuntime(serverless) {
-        if (serverless.service.provider.runtime === 'provided') {
+    checkCompatibleRuntime() {
+        if (this.serverless.service.provider.runtime === 'provided') {
             throw new Error('Bref 1.0 layers are not compatible with the "provided" runtime. To upgrade to Bref 1.0, you have to switch to "provided.al2" in serverless.yml. More details here: https://bref.sh/docs/news/01-bref-1.0.html#amazon-linux-2');
         }
-        for (const [name, f] of Object.entries(serverless.service.functions)) {
+        for (const [name, f] of Object.entries(this.serverless.service.functions)) {
             if (f.runtime === 'provided') {
                 throw new Error(`Bref 1.0 layers are not compatible with the "provided" runtime. To upgrade to Bref 1.0, you have to switch to "provided.al2" in serverless.yml for the function "${name}". More details here: https://bref.sh/docs/news/01-bref-1.0.html#amazon-linux-2`);
             }
         }
     }
+
+    async createVendorZip() {
+        if(! this.serverless.service.custom.separateVendor) {
+            return;
+        }
+
+        const vendorZipHash = await this.createZipFile();
+        this.newVendorZipName = vendorZipHash + '.zip';
+
+        let excludes = this.serverless.service.package.exclude;
+        if(excludes.indexOf('vendor/**') === -1) {
+            excludes[excludes.length] = 'vendor/**';
+        }
+
+        //let iamRoleStatements = this.serverless.service.provider.iamRoleStatements;
+        const roleDetails = {
+            'Effect': 'Allow',
+            'Action': [
+                's3:GetObject',
+            ],
+            'Resource': [
+                {
+                    'Fn::Join': [
+                        '',
+                        [
+                            'arn:aws:s3:::',
+                            {
+                                'Ref': 'ServerlessDeploymentBucket'
+                            },
+                            '/',
+                            this.stripSlashes(this.provider.getDeploymentPrefix() + '/vendors/*')
+                        ]
+                    ]
+                }
+            ]
+        };
+
+        let resources = this.serverless.service.provider.coreCloudFormationTemplate.Resources;
+        resources[resources.length] = {
+            'ServerlessDeploymentBucketLambdaAccessPolicy': roleDetails
+        };
```

Comment:
It shouldn't replace it (I hope)... Basically it just appends an item to the cloud formation template resources, but I'm not 100% sure. I had no case where I could have tested this scenario. I tried doing this with the iamRoleStatements at first, but they were already parsed at that moment and I couldn't get the name of the deployment bucket and prefix without using references, but references are somehow not working after parsing the config.

If there's a way so I can use a reference outside of a `Fn::Join` I'd love to get some input/help here. 😅

Reply:
Yep I see what you mean but I don't see how this can work actually, the resource seems invalid because incorrectly formatted 🤔 

How about adding the permission to [iamRoleStatements](https://www.serverless.com/framework/docs/providers/aws/guide/iam/):

```js
this.serverless.service.provider.iamRoleStatements = this.serverless.service.provider.iamRoleStatements ? this.serverless.service.provider.iamRoleStatements : [];

this.serverless.service.provider.iamRoleStatements.push({
            Effect: 'Allow',
            Action: [
                's3:GetObject',
            ],
            Resource: [
                {
                    'Fn::Join': [
                        '',
                        [
                            'arn:aws:s3:::',
                            {
                                'Ref': 'ServerlessDeploymentBucket'
                            },
                            '/',
                            this.stripSlashes(this.provider.getDeploymentPrefix() + '/vendors/*')
                        ]
                    ]
                }
            ]
});
```

(haven't tested to be 100% transparent)

Reply:
Wait wait, now I'm wondering: at the `after:aws:deploy:deploy:createStack` step isn't the serverless config already parsed to produce the CloudFormation template?

Now I'm wondering if the vendor exclusion (and what I suggest in my comment above) are taken into account at all 🤔 

It may be worth moving those changes into the `package:setupProviderConfiguration` hook?

Reply:
Yes, that sounds like a good plan, thank you. I'll add it right away. I had something like that in an earlier version but due to the problems with the `ServerlessDeploymentBucket` I removed it again. (That's why there was still a comment left for the `iamRoleStatements` selection.

I think this may be IMHO the best solution to add the rights for the s3 bucket. 👍

Reply:
> Wait wait, now I'm wondering: at the `after:aws:deploy:deploy:createStack` step isn't the serverless config already parsed to produce the CloudFormation template?

Yes, it's already parsed, that's why I added it as a new element to the parsed template previously.

> Now I'm wondering if the vendor exclusion (and what I suggest in my comment above) are taken into account at all 🤔

In my tests this worked fine, I didn't have vendor exclusion in my serverless.yml file while running with separation enabled and the vendor folder was excluded fine.

> It may be worth moving those changes into the `package:setupProviderConfiguration` hook?

Ok I'll see if they are working inside that hook.

Reply:
> It may be worth moving those changes into the package:setupProviderConfiguration hook?
@mnapoli we are already in that hook with this code. 😅

Reply:
Oh my bad!

If it works, then it works! 😅